### PR TITLE
Cull service UI 

### PR DESCRIPTION
**Pull request by dvigneshwer**
A monitoring page is started which provides a text field to enter the time out value and start cull service.

The page also displays all the notebook servers with the last activity and removes them after being culled.

This feature is an extended version of cull service with a UI to start, monitor and stop cull service.

**Before cull service started :**

image

**After cull service started :** 

image

**Prompt to enter timer value, on click of cull service button**

image

### CODE REVIEW

File: examples/cull-idle/Cull_service_UI/Readme.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<h2><strong>Cull Service UI Monitoring</strong></h2>
```
Comment (minrk) [inline examples/cull-idle/Cull_service_UI/Readme.md]: The filename suggests this is a markdown file, but this is HTML. Did you commit the wrong file?

Comment (dvigneshwer) [inline examples/cull-idle/Cull_service_UI/Readme.md]: @minrk I worte the html syntax in the markdown file as github renders it properly in the readme page, I can change it to markdown syntax if you suggest.

Comment (minrk) [inline examples/cull-idle/Cull_service_UI/Readme.md]: I think that would be preferable, as it is the norm for markdown documentation.

File: examples/cull-idle/Cull_service_UI/web_service.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+from dateutil.parser import parse as parse_date 
+import json
+
+define("port", default=8888, help="run on the given port", type=int)
+define("debug", default=False, help="run in debug mode")
+define("static_path", default="./", help="Content of the template")
+define('url', default='https://127.0.0.1:443/hub', help="The JupyterHub API URL")
+
+# Values to be changed
+api_token="c4297f02c25b49c2a63c9b102b24cc51"
```
Comment (minrk) [inline examples/cull-idle/Cull_service_UI/web_service.py]: it's a good idea to get the API token from an environment variable. See the existing cull_idle example for reference. The same goes for the other constants below - they can come from the options above, and don't need to be defined as constants here.

File: examples/cull-idle/Cull_service_UI/web_service.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+cull_time=1000
+
+# Constant variables
+clients = [] 
+session_info={}
+time_out_val=None
+
+# Renders the index page
+class MainHandler(tornado.web.RequestHandler):
+	@coroutine
```
Comment (minrk) [inline examples/cull-idle/Cull_service_UI/web_service.py]: Looks like you have indented with tabs. The common convention in Python is to indent with four spaces. What editor are you using? You can generally configure it to indent with different things for different file types.

File: examples/cull-idle/Cull_service_UI/js/script.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+// creating a new socket connection
+var socket = new WebSocket("ws://127.0.0.1:8888/ws");
+
+// socket function to interact with the web server
+socket.onopen = function() {
+	console.log("connected to web socket");
+}
+
+socket.onmessage = function(message) {
+	console.log("Recieving :-" + message.data);
```
Comment (minrk) [inline examples/cull-idle/Cull_service_UI/js/script.js]: Looks like we have tab-indentation here as well. The convention in JavaScript is two spaces, though I won't object to four if that's easier for you.

**Comment by minrk** [comment]
It looks like the cull service only runs when there are websocket connections. I don't think this is the desired behavior. What we probably want is the cull to run all the time, and for pages to connect to get updates. Viewing the status page should not affect whether culling is happening or not.

File: examples/cull-idle/Cull_service_UI/Readme.md (hunk 1 above)
Comment (dvigneshwer) [inline examples/cull-idle/Cull_service_UI/Readme.md]: Yeah will change them, sorry for the inconvenience.

File: examples/cull-idle/Cull_service_UI/js/script.js (hunk 4 above)
Comment (dvigneshwer) [inline examples/cull-idle/Cull_service_UI/js/script.js]: I use tab width 4 in sublime, find it more structured with tab indentation.

File: examples/cull-idle/Cull_service_UI/web_service.py (hunk 2 above)
Comment (dvigneshwer) [inline examples/cull-idle/Cull_service_UI/web_service.py]: Sorry was being lazy and excited to send the PR, bad from my side will change it. I will use the environment var itself as done in the cull_service.py cod by you.

File: examples/cull-idle/Cull_service_UI/web_service.py (hunk 3 above)
Comment (dvigneshwer) [inline examples/cull-idle/Cull_service_UI/web_service.py]: I use tab width 4 in sublime (shows tab size 4 in UI)

Comment (minrk) [inline examples/cull-idle/Cull_service_UI/web_service.py]: That's fine, but it's good to be consistent within a given project. The norm in Python (and used throughout the rest of this repo) is to use four spaces. When contributing to projects, it is a good idea to follow the standards already set there. Jupyter projects aren't as strict as many, but we do require the basics of python style like four-space indentation.

Comment (dvigneshwer) [inline examples/cull-idle/Cull_service_UI/web_service.py]: Thanks for the review will change it to four space indentation and let you know. I wasn't aware of the above fact.

**Comment by dvigneshwer** [comment]
Yeah need to tweak that, will make that modification and let you know.

**Comment by willingc** [comment]
@dvigneshwer I'm going to reopen this pull request so that you may submit additional changes here.

When submitting changes to a PR, it is possible to push the changes to the same branch and it will automatically update the PR. For example, if you push changes to your ``master`` branch, they will update here.

While you may submit a new PR, it is more common practice in open source to update an existing PR with changes since it keeps all review comments and changes together in the repo history. Please let me know if you have any questions on submitting the changes to this PR.

**Comment by dvigneshwer** [comment]
Okay, thanks @willingc will push in the same branch.

**Comment by willingc** [comment]
Hi @dvigneshwer, I'm going to close this PR. Thanks for the effort and we hope you contribute to Jupyter in the future.

**Pull request closed without merging**